### PR TITLE
feature #4024 added a `list-files` command

### PR DESCRIPTION
**Pull request by clxmstaab**
as discussed in https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4024 this PR proposes a `list-files` command, which from a user-perspective can be used to utilize `xargs` to get some basic parallelization into php-cs-fixer.

this is especially interessting for CI build cases, in which the php-cs-fixer builtin caching mechanism doesn't work (e.g. because git does use the correct file modification stamps on clone or the CI build is running within a readonly environment).

example usage:
```
vendor/bin/php-cs-fixer list-files | xargs -n 10 -P 8 vendor/bin/php-cs-fixer fix --config=.php_cs.dist --using-cache=no
```
`-n` defines how many files a single subprocess process
`-P` defines how many subprocesses the shell is allowed to spawn for parallel processing (usually similar to the number of CPUs your system has)

see xargs help page: https://wiki.ubuntuusers.de/xargs/

as can be seen in https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4024#issuecomment-750004737 you might get a perf boost by a factor of 3-6x depending on the number of files beeing fixed, number of CPUs etc.

this isn't the most elegant way of how parallelzation can be applied on the php-cs-fixer case, but its a really simple one which works with very simple means and therefore has a rather low entry barrier.
you don't need fancy php extension or any rather complicated setup. simple bash means are enought to get a decent perf boost.
also its pretty simple from php-cs-fixer perspective to keep the maintenance burden as low as possible.

in case you would consider this change acceptable, I am willing to provide automated tests, docs and everything else required and missing atm. 

any opinions?

### CODE REVIEW

File: src/Console/Application.php
Comment (clxmstaab) [inline src/Console/Application.php:56]: the command is intentionally not named `ListCommand` as this would override the symfony builtin `ListCommand` and cause several unexpected side-effects

**Comment by keradus** [comment]
> in which the php-cs-fixer builtin caching mechanism doesn't work (e.g. because git does use the correct file modification stamps on clone or the CI build is running within a readonly environment)

Just a note, Cache mechanism relies on checksum of file, not it's timestamp, so this is not the case.
Indeed, if you have readonly mechanism, then you cannot use the cache easily

File: src/Console/Application.php
Comment (keradus) [inline src/Console/Application.php:15]: I'm not sure whether the case of not being able to use the cache is strong enough to have build-in support for workaround, that we need to maintain in the codebase.
what do you think, @localheinz @julienfalque @kubawerlos @SpacePossum ?

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [inline src/Console/Command/ListFilesCommand.php:75]: maybe we should have a switch to filter-out files that do not require a fixing (based on cache) ?

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [inline src/Console/Command/ListFilesCommand.php:62]: if we would decide to have this command, we should document it's purpose and usage

File: src/Console/Application.php
Comment (localheinz) [reply src/Console/Application.php:56]: What do you think about adding the command in alphabetical order?

File: src/Console/Command/ListFilesCommand.php
Comment (localheinz) [reply src/Console/Command/ListFilesCommand.php:75]: 👍 

Sounds good!

File: src/Console/Application.php
Comment (localheinz) [reply src/Console/Application.php:15]: @keradus 

On one hand, if anyone cannot use the cache in a CI environment, I would argue that maybe something is wrong with the CI setup. On the other hand, this command is small, so I would be fine with adding it!

File: src/Console/Command/ListFilesCommand.php
Comment (localheinz) [inline src/Console/Command/ListFilesCommand.php:40]:
```suggestion


```

File: src/Console/Command/ListFilesCommand.php
Comment (localheinz) [inline src/Console/Command/ListFilesCommand.php:86]:
```suggestion
```

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [inline src/Console/Command/ListFilesCommand.php:103]: do we need real path here?

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [inline src/Console/Command/ListFilesCommand.php:36]: please add missing tests

**Comment by mfn** [comment]
One use-case which wasn't mentioned, in which actually triggered my curiosity in finding a solution speed up non-cached runs: changing the rules.

Doesn't sound like much, but: if you consider a large code base where, as I demonstrated in https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4024#issuecomment-750004737 , a run takes 1m40s: that's no found to _try out rules_ and analyze the impact. In fact I was going through lots-o-rules and couldn't bear it after 2 or 3 changes, so I came up with this hack with xargs, because ~30s is acceptable.

Thanks!

File: src/Console/Command/ListFilesCommand.php
Comment (clxmstaab) [reply src/Console/Command/ListFilesCommand.php:103]: I am primarily use the `realpath` to normalize directory separators.

File: src/Console/Application.php
Comment (clxmstaab) [reply src/Console/Application.php:15]: in case I get your "go", I will add tests etc.

File: src/Console/Command/ListFilesCommand.php
Comment (clxmstaab) [reply src/Console/Command/ListFilesCommand.php:75]: I don't know the internals of php-cs enough to implement the mentioned switch. could you give me a hint of what you are talking about, or whether there is another code-site where I can get inspiration from?

File: src/Console/Command/ListFilesCommand.php
Comment (clxmstaab) [reply src/Console/Command/ListFilesCommand.php:62]: yeah,.. in the first place I just copy/pasted the fix command and reduced it, to what we can see atm.

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [reply src/Console/Command/ListFilesCommand.php:100]: yet it increase the length of path. not sure if really needed, to be honest

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [reply src/Console/Command/ListFilesCommand.php:75]: basically if the file has no changes, we do not need to fix it, thus we may not want to list it

you may want to take a look here:
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/4874565ef708da70481a13a21d0c28b39ef53008/src/Runner/Runner.php#L121
- https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/4874565ef708da70481a13a21d0c28b39ef53008/src/Runner/FileFilterIterator.php

File: src/Console/Application.php
Comment (SpacePossum) [reply src/Console/Application.php:15]: The `xargs` solution might be the best we got for a while for multi threading (until we get (better) multi threading for PHP ootb/better 3rd party).
 
If supporting all it takes is the new command `as is` + additional the cache check than I would be in favor (with the cache check support added, documentation, etc. all is on par with project requirements)

**Comment by keradus** [comment]
@clxmstaab , please ping me if you would need another round of review!
eager to merge this feature when it's ready ;)

**Comment by mfn** [comment]
> If for supporting it all it takes is the new command as is + additional the cache check
…

> maybe we should have a switch to filter-out files that do not require a fixing (based on cache) ?
…

My modest feedback: I would like to see _this PR_ focus only on that one feature: return the list of files as it would be found by the finder.

Let the connection to the cache be a follow-up change, if it's strongly desired.

The feature will be useful "as is" and it keeps the scope / review process easier.

My 2c.

File: src/Console/Command/ListFilesCommand.php
Comment (clxmstaab) [reply src/Console/Command/ListFilesCommand.php:75]: for the use-case I foresee (see the referenced issue), the cache will be empty in most situations.
IMO we could add support for using the cache later, in case we get use-cases which benefit from it.

since this new option would add additional complexity and takes time, I would initially add the feature without a cache option - so we can get the basic feature into the tool earlier.

File: tests/Console/Command/test-project/.php_cs
Comment (clxmstaab) [inline tests/Console/Command/test-project/.php_cs:1]: I don't like that I needed to add this test assets. maybe someone can give me a hint, what I could use instead?

File: src/Console/Command/ListFilesCommand.php
Comment (clxmstaab) [inline src/Console/Command/ListFilesCommand.php:86]: while real-world testing, I got 'sometimes' a error

```
...
'tests/Tokenizer/TransformersTest.php'
'tests/ToolInfoTest.php'
'tests/UtilsTest.php'
'tests/WhitespacesFixerConfigTest.php'
'tests/WordMatcherTest.php'
PHP Fatal error:  Uncaught Error: Call to undefined method SplFileInfo::getRelativePathname() in /mnt/c/dvl/GitHub/PHP-CS-Fixer/src/Console/Command/ListFilesCommand.php:86
Stack trace:
#0 /mnt/c/dvl/GitHub/PHP-CS-Fixer/vendor/symfony/console/Command/Command.php(255): PhpCsFixer\Console\Command\ListFilesCommand->execute()
#1 /mnt/c/dvl/GitHub/PHP-CS-Fixer/vendor/symfony/console/Application.php(971): Symfony\Component\Console\Command\Command->run()
#2 /mnt/c/dvl/GitHub/PHP-CS-Fixer/vendor/symfony/console/Application.php(290): Symfony\Component\Console\Application->doRunCommand()
#3 /mnt/c/dvl/GitHub/PHP-CS-Fixer/src/Console/Application.php(93): Symfony\Component\Console\Application->doRun()
#4 /mnt/c/dvl/GitHub/PHP-CS-Fixer/vendor/symfony/console/Application.php(166): PhpCsFixer\Console\Application->doRun()
#5 /mnt/c/dvl/GitHub/PHP-CS-Fixer/php-cs-fixer(102): Symfony\Component\Console\Application->run()
#6 {main}
  thrown in /mnt/c/dvl/GitHub/PHP-CS-Fixer/src/Console/Command/ListFilesCommand.php on line 86
```

its strange, since it works for nearly all files, but only a few error.
does the sfFinder sometimes return a `\SplFileInfo` and sometimes a `Symfony\Component\Finder\SplFileInfo`?
(`getRelativePathname` is only defined on the symfony-SplFileInfo class)

**Comment by clxmstaab** [comment]
any feedback on the current status @keradus ?

File: tests/Console/Command/test-project/.php_cs
Comment (keradus) [reply tests/Console/Command/test-project/.php_cs:1]: you can move it to `tests/Fixtures`, and then it gonna be excluded everywhere where needed

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [reply src/Console/Command/ListFilesCommand.php:86]: that's probably not the only place, but looks like Finder can return built-in `\SplFileInfo` too
https://github.com/symfony/finder/blob/5.x/Finder.php#L660

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [inline src/Console/Command/ListFilesCommand.php:61]:
```suggestion
                    new InputOption('config', '', InputOption::VALUE_REQUIRED, 'The path to a .php-cs-fixer.php file.'),
```

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [reply src/Console/Command/ListFilesCommand.php:86]: can you reproduce the issue with unit test?

File: tests/Console/Command/test-project/.php_cs
Comment (keradus) [inline tests/Console/Command/test-project/.php_cs:7]: maybe put sth excluded too, to verify that we take all "in" and "out", not only "in"

**Comment by keradus** [comment]
looking good, just some minors in CodeReview, CI should be fixed, and I'm really curious about that relative path

File: src/Console/Command/ListFilesCommand.php
Comment (keradus) [reply src/Console/Command/ListFilesCommand.php:86]: worst case scenario, simply check if object we got implements sf-file or \file and use one path or another

File: src/Console/Application.php
Comment (keradus) [inline src/Console/Application.php:18]: please add example usage command in the docs

**Comment by clxmstaab** [comment]
I think this should be good to go.

the github actions are pretty slow right now, so the results will show whether I am right.
the static analysis errors looks unrelated to me.

> does the `sfFinder` sometimes return a `\SplFileInfo` and sometimes a `Symfony\Component\Finder\SplFileInfo`?
> (`getRelativePathname` is only defined on the symfony-SplFileInfo class)


I did some path gymnastics and I think this resolved the Problem. the new version works with what is available in `\SplFileInfo`, and thats ok for the `sfSplFileInfo` because it `extends \SplFileInfo`.

feedback welcome.

**Comment by clxmstaab** [comment]
looking at CI, it looks to me that the errors are unrelated

File: tests/Fixtures/ListFilesTest/.php-cs-fixer.php
Comment (keradus) [inline tests/Fixtures/ListFilesTest/.php-cs-fixer.php:12]: please fix the deprecation travis is reporting: https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5390/checks?check_run_id=2351718634#step:11:523

File: tests/Fixtures/ListFilesTest/.php-cs-fixer.php
Comment (keradus) [inline tests/Fixtures/ListFilesTest/.php-cs-fixer.php:15]:
```suggestion
        '@Symfony' => true,
```

File: doc/usage.rst
Comment (keradus) [inline doc/usage.rst:148]: would it make sense to count the files and provide `files/cpu` to single job?

**Comment by keradus** [comment]
> looking at CI, it looks to me that the errors are unrelated

i merged the [fix](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5629), can you rebase to prove it was only related to it?
(eg that one deprecation i found looks actually related to PR)

File: doc/usage.rst
Comment (keradus) [inline doc/usage.rst:136]:
```suggestion
The ``--config`` option can be used, like in the ``fix`` command, to tell from which path a config file should be loaded
```

File: doc/usage.rst
Comment (keradus) [reply doc/usage.rst:148]: to provide multiple files, you need to specify config file (ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/4279#issuecomment-820615223  )

Comment (clxmstaab) [reply doc/usage.rst]: > would it make sense to count the files and provide =$files/$cpu to single job?

not sure what you mean or what you are after?

> to provide multiple files to fix command


hmm does this mean we should make the config a required argument instead of a option?

**Comment by clxmstaab** [comment]
> i merged the [fix](https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5629), can you rebase to prove it was only related to it?

it seems this works now, thx

**Comment by keradus** [comment]
as I was afraid, after the fix for main branch was merged, this PR is still red - Windows failing, 5.6 failing, fabbot failing. can you take a look?

**Comment by clxmstaab** [comment]
fixed the windows builds, thx for the headsup.

the macos buid failure seems unrelated

File: doc/usage.rst
Comment (keradus) [inline doc/usage.rst:140]:
```suggestion
    $ php php-cs-fixer.phar list-files --config=.php-cs-fixer.dist.php
```

File: doc/usage.rst
Comment (keradus) [inline doc/usage.rst:150]:
```suggestion
    $ php php-cs-fixer.phar list-files --config=.php-cs-fixer.dist.php | xargs -n 10 -P 8 php-cs-fixer fix --config=.php-cs-fixer.dist.php
```

**Review by keradus — APPROVED** [review]

**Comment by keradus** [comment]
Really nice work, @clxmstaab !

**Comment by keradus** [comment]
Thank you @clxmstaab.

**Pull request merged**
